api/opds.py

```diff
@@ -81,6 +81,14 @@ def __init__(self, lane,
         self.hidden_content_types = hidden_content_types
         self.test_mode = test_mode
 
+    def is_work_entry_complete(self, work):
+        """Return a boolean value indicating whether the work's OPDS catalog entry is complete.
```

Comment:
I can see a couple of problems with this. Most importantly, this will return true if this Annotator is prepared to generate a complete OPDS entry for _any_ work. It should only return true if the Annotator is prepared to generate a complete OPDS entry for _this_ work. So the code should look like:

```
return any(work in x for x in self.active_loans_by_work, self.active_holds_by_work, self.active_fulfillments_by_work)
```

Second, I don't think "complete catalog entry" is the right concept here. Section 5.1.2 of the OPDS spec says a complete catalog entry must "include all metadata elements".  I don't understand how an active loan/hold/fulfillment makes the difference here. An acquisition link isn't a metadata element.

I think it would be better to change the link from rel="alternate" to rel="self" across the board -- all of these OPDS entries are complete, and the use of "self" instead of "alternate" to indicate a complete entry makes sense as a convention.

Assuming it doesn't cause problems with existing clients, is there any reason not to _always_ use rel="self" here?

Reply:
This is a very good question, indeed. This issue was raised on behalf of Saad, who was working on integrating CM OPDS feeds into Thorium. Since all the catalog entries can be considered "complete" according to the spec, I think it should be sufficient and absolutely right. I'll follow up with Saad and return with feedback.

Reply:
Thinking it out, the distinction I'm seeing is between an OPDS entry which is generic, not customized with special acquisition links based on the authenticated patron (e.g. because it's in a big feed that came from a cache), and an OPDS entry that _is_ customized (because of your recent change to the `permalink` endpoint).

This is a good distinction to make but it doesn't have a basis in the spec. I can see the appeal of using rel="alternate" in the big feed and rel="self" in the permalink for this reason.